daemon/api_apps.go

```diff
@@ -223,10 +225,97 @@ func getLogs(c *Command, r *http.Request, user *auth.UserState) Response {
 
 var servicestateControl = servicestate.Control
 
-func postApps(c *Command, r *http.Request, user *auth.UserState) Response {
+func decodeServiceInstruction(body io.ReadCloser, u *user.User) (*servicestate.Instruction, error) {
 	var inst servicestate.Instruction
-	decoder := json.NewDecoder(r.Body)
+	decoder := json.NewDecoder(body)
 	if err := decoder.Decode(&inst); err != nil {
+		return nil, err
+	}
+
+	// If the list of users to affect is simply 'user', then lets fixup that
+	// by the username of the invoker. If the list contains 'users' then we simply
+	// mean all users.
+	var updated []string
+	for _, name := range inst.ServicesOfUsers {
+		if name == "users" {
+			updated = nil
+			break
+		}
+
+		if name == "user" {
+			// Reference to the current user, if the user is root this not valid
+			// as root users should only use 'users' for all, or provide a list
+			// of user names
+			if u == nil || u.Uid == "0" {
+				return nil, fmt.Errorf("cannot use \"user\" in a root user context for service operations")
+			}
+			updated = append(updated, u.Username)
+		} else {
+			updated = append(updated, name)
+		}
+	}
+	inst.ServicesOfUsers = updated
+	return &inst, nil
+}
+
+func hasUserService(apps []*snap.AppInfo) bool {
+	for _, app := range apps {
+		if app.IsService() && app.DaemonScope == snap.UserDaemon {
+			return true
+		}
+	}
+	return false
+}
+
+var systemUserFromRequest = func(r *http.Request) (*user.User, error) {
+	uid, err := uidFromRequest(r)
+	if err != nil {
+		return nil, err
+	}
+	return user.LookupId(strconv.Itoa(int(uid)))
+}
+
+func validateScopeAgainstApps(inst *servicestate.Instruction, u *user.User, apps []*snap.AppInfo) error {
+	// Set default scopes if not provided
+	if len(inst.Scope) == 0 {
+		// If root is making this request, implied scopes are all
+		if u == nil || u.Uid == "0" {
+			inst.Scope = append(inst.Scope, "system", "user")
```

Comment:
Hmm @pedronis what is your opinion here. I think it could make sense for a root-user not to by default restart user-daemons, but instead initially, only scope to system services (unless user scope is requested)

Reply:
we should have a chat about this